src/tox/config/source/discover.py

```diff
@@ -27,6 +27,9 @@ def discover_source(config_file: Path | None, root_dir: Path | None) -> Source:
         if src is None:
             src = _create_default_source(root_dir)
     else:
+        if config_file.is_dir():
+            if (config_file / "tox.ini").is_file():
+                config_file = config_file / "tox.ini"
         src = _load_exact_source(config_file)
```

Comment:
This needs to be generic, so does not belong here. Needs to work for other configuration sources, not just `tox.ini`.  Take a look here for other sources:

https://github.com/tox-dev/tox/blob/rewrite/src/tox/config/source/setup_cfg.py
https://github.com/tox-dev/tox/blob/rewrite/src/tox/config/source/legacy_toml.py

Reply:
TBH, I do not want to spend time trying to open N configuration filenames and see if they include tox section when a folder is gives, especially "legacy" ones. As v4 is a major version, I think we safely drop some from the list.

WDYT? Or maybe you have a suggestion about how to implement it and reuse tox own abilities?

If the argument is a folder, we could just do a pushd(folder) and run tox from there, without -c? Is that doable?

Reply:
> TBH, I do not want to spend time trying to open N configuration filenames and see if they include tox section when a folder is gives, especially "legacy" ones. As v4 is a major version, I think we safely drop some from the list.

Strong disagree.

Reply:
I'd imagine we'd repeat this loop here with the config file name passed is a directory. See how you can get the filename to fill in by using `s.FILENAME`.